rfcs/2020-11-01-remap-metrics-support.md

```diff
@@ -0,0 +1,219 @@
+# RFC 3550 - 2020-11-01 - Support metrics with the new 'remap' transform
+
+It would be useful to add the ability for the Remap language to work with metric
+data as well as with log data.
+
+## Scope
+
+The aim of the Remap language is to remain simple and yet provide the power to
+allow the user to replace most of their tranform configuration with a simple
+Remap script. Any complexity beyond a certain level is possible, but requires
+the `Lua` or `Wasm` transforms.
+
+Given this, the scope of this RFC is not to enhance the Remap language so it
+can work comprehensively with metric data. The main aim will be to allow the
+user to make simple changes to the metadata surrounding the metric.
+
+
+## Motivation
+
+Currently there are some simple transforms that allow manipulating the metadata
+of metric data - `add_tags`, `remove_tags`, `rename_tags`. These transforms
+are useful, but limited. They cannot take into account any additional data around
+the metric to determine how to process the tags.
+
+
+## Internal Proposal
+
+Currently to access data via Paths, `Remap` requires an object that implements 
+the `remap::Object` trait. Currently, the implementation only works for log 
+events.
```

Comment:
To add some more context:

We currently implement `impl remap::Object for vector::Event`. This exposes the `insert`, `find`, `paths` and `remove` functions.

To support both log events and metric events, I think the best course of action is to actually implement `impl remap::Object` individually for `LogEvent` and `MetricEvent`, _not_ for the enum `Event`, the way it is done right now.

This gives us better separation of the two implementations.

When, instead of doing `execute(&mut event)`, we'd do `execute(event.as_mut_log()` and `execute(event.as_mut_metric()`, depending on if we're dealing with a log or metric.

Reply:
Yeah, that feels cleaner, but I think that might depend on how #4605 pans out? That seems to want access to both metric and log sides. It probably needs a bit of thought before we come up with a final solution. I think it is an implementation detail that shouldn't distract too much.

Reply:
I like this idea @JeanMertz and it will help us avoid confusion later.